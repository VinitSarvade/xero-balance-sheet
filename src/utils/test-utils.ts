import {
  format as prettyFormat,
  plugins as prettyFormatPlugins,
} from "pretty-format";
import { http, HttpResponse } from "msw";
import { BalanceSheetReport } from "@/modules/xero/xeroReport.dto";

export function format(value: any) {
  return prettyFormat(value, {
    escapeRegex: true,
    indent: 2,
    plugins: [
      prettyFormatPlugins.ReactTestComponent,
      prettyFormatPlugins.ReactElement,
      prettyFormatPlugins.DOMElement,
      prettyFormatPlugins.DOMCollection,
      prettyFormatPlugins.Immutable,
      prettyFormatPlugins.AsymmetricMatcher,
    ],
    printFunctionName: false,
  });
}

export const BalanceSheetMock = {
  ReportID: "report-id",
  ReportName: "Balance Sheet",
  ReportDate: "2022-01-01",
  ReportTitles: ["Balance Sheet", "Test Company"],
  Rows: [
    {
      RowType: "Header",
      Cells: [{ Value: "" }, { Value: "2023" }, { Value: "2024" }],
    },
    {
      RowType: "Section",
      Title: "Assets",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Bank",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "My Bank Account" },
            { Value: "126.70" },
            { Value: "99.60" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business" },
            { Value: "92911.00" },
            { Value: "92911.00" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Sample Business 1" },
            { Value: "11039.00" },
            { Value: "11039.00" },
          ],
        },
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Bank" },
            { Value: "104076.70" },
            { Value: "104049.60" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "Liabilities",
      Rows: [],
    },
    {
      RowType: "Section",
      Title: "Current Liabilities",
      Rows: [
        {
          RowType: "Row",
          Cells: [
            { Value: "Car Loan" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
    {
      RowType: "Section",
      Title: "",
      Rows: [
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Current Liabilities" },
            { Value: "8000.00" },
            { Value: "8000.00" },
          ],
        },
      ],
    },
  ],
} as BalanceSheetReport;

export function getHandlers() {
  return [
    http.get(`${process.env.BASE_URL}/api/xero/balance-sheet`, () => {
      return HttpResponse.json(BalanceSheetMock);
    }),
  ];
}
