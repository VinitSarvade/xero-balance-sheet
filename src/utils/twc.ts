import { createTwc } from "react-twc";
import { cn } from "./cn";

export const twc = createTwc({ compose: cn });
