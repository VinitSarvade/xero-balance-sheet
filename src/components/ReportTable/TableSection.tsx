import { SectionRow } from "@/modules/xero/xero.module";
import { cn } from "@/utils/cn";
import TableSectionRow from "./TableSectionRow";

interface TableSectionProps {
  section: SectionRow;
}

export default function TableSection({ section }: TableSectionProps) {
  const hasEntries = section.Rows.length > 0;
  const hasTitle = section.Title.length > 0;

  const isPrimarySectionTitle =
    section.Title === "Assets" || section.Title === "Liabilities";
  const isSubSectionTitleRow = !isPrimarySectionTitle && hasEntries && hasTitle;

  return (
    <>
      <tr key={`section-${section.Title}`}>
        <td colSpan={3} className="py-2">
          <div
            className={cn("text-xl font-semibold", {
              "text-2xl": isPrimarySectionTitle,
              "pl-4": isSubSectionTitleRow,
            })}
            data-testid="section-title"
          >
            {section.Title}
          </div>
        </td>
      </tr>

      {section.Rows.map((row) => (
        <TableSectionRow
          key={`row-${row.RowType}-${section.RowType}-${row.Cells[0].Value}`}
          row={row}
          section={section}
        />
      ))}
    </>
  );
}
