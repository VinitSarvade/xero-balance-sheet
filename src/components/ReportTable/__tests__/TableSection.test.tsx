import { expect, describe, it } from "bun:test";
import { render } from "@testing-library/react";
import { SectionRow } from "@/modules/xero/xeroReport.dto";
import TableSection from "../TableSection";
import { format } from "@/utils/test-utils";

const section: SectionRow = {
  Title: "Assets",
  RowType: "Section",
  Rows: [
    {
      RowType: "Row",
      Cells: [{ Value: "Cash" }, { Value: "100" }, { Value: "120" }],
    },
    {
      RowType: "Row",
      Cells: [{ Value: "Checking Account" }, { Value: "50" }, { Value: "100" }],
    },
    {
      RowType: "Row",
      Cells: [{ Value: "Savings Account" }, { Value: "50" }, { Value: "500" }],
    },
    {
      RowType: "SummaryRow",
      Cells: [{ Value: "Total Assets" }, { Value: "200" }, { Value: "720" }],
    },
  ],
};

describe("TableSection", () => {
  it("renders primary section title correctly", () => {
    const section: SectionRow = {
      Title: "Assets",
      RowType: "Section",
      Rows: [
        {
          RowType: "Row",
          Cells: [{ Value: "Cash" }, { Value: "100" }, { Value: "120" }],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Checking Account" },
            { Value: "50" },
            { Value: "100" },
          ],
        },
        {
          RowType: "Row",
          Cells: [
            { Value: "Savings Account" },
            { Value: "50" },
            { Value: "500" },
          ],
        },
        {
          RowType: "SummaryRow",
          Cells: [
            { Value: "Total Assets" },
            { Value: "200" },
            { Value: "720" },
          ],
        },
      ],
    };
    const { getByText } = render(
      <table>
        <tbody>
          <TableSection section={section} />
        </tbody>
      </table>,
    );
    const sectionTitle = getByText("Assets");
    expect(sectionTitle).toBeTruthy();
    expect(sectionTitle.tagName).toBe("DIV");
    expect(sectionTitle.className).toContain("text-2xl");
    expect(sectionTitle.className).toContain("font-semibold");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <table>
        <tbody>
          <TableSection section={section} />
        </tbody>
      </table>,
    );

    expect(format(asFragment())).toMatchSnapshot();
  });

  it("renders sub-section title row correctly", () => {
    const section: SectionRow = {
      Title: "Current Assets",
      RowType: "Section",
      Rows: [
        {
          RowType: "Row",
          Cells: [{ Value: "Cash" }, { Value: "100" }, { Value: "100" }],
        },
      ],
    };
    const { getByText } = render(
      <table>
        <tbody>
          <TableSection section={section} />
        </tbody>
      </table>,
    );
    const sectionTitle = getByText("Current Assets");
    expect(sectionTitle).toBeTruthy();
    expect(sectionTitle.tagName).toBe("DIV");
    expect(sectionTitle.className).toContain("text-xl");
    expect(sectionTitle.className).toContain("font-semibold");
    expect(sectionTitle.className).toContain("pl-4");
  });

  it("renders no section title when title is empty", () => {
    const section: SectionRow = {
      Title: "",
      RowType: "Section",
      Rows: [
        {
          RowType: "Row",
          Cells: [{ Value: "Cash" }, { Value: "100" }, { Value: "100" }],
        },
      ],
    };
    const { findByTestId } = render(
      <table>
        <tbody>
          <TableSection section={section} />
        </tbody>
      </table>,
    );
    const sectionTitle = findByTestId("section-title");
    expect(sectionTitle).toBeEmpty();
  });

  it("renders no rows when Rows array is empty", () => {
    const section: SectionRow = {
      Title: "Current Assets",
      RowType: "Section",
      Rows: [],
    };
    const { queryAllByRole } = render(
      <table>
        <tbody>
          <TableSection section={section} />
        </tbody>
      </table>,
    );
    const rows = queryAllByRole("row");
    expect(rows).toHaveLength(1);
  });
});
