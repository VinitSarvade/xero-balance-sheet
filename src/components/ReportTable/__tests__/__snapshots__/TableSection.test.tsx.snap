// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSection matches snapshot 1`] = `
"<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="py-2"
          colspan="3"
        >
          <div
            class="font-semibold text-2xl"
            data-testid="section-title"
          >
            Assets
          </div>
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="px-2 py-1.5 pl-4"
        >
          Cash
        </td>
        <td
          class="px-2 text-right"
        >
          ₹100.00
        </td>
        <td
          class="px-2 text-right"
        >
          ₹120.00
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="px-2 py-1.5 pl-4"
        >
          Checking Account
        </td>
        <td
          class="px-2 text-right"
        >
          ₹50.00
        </td>
        <td
          class="px-2 text-right"
        >
          ₹100.00
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class="px-2 py-1.5 pl-4"
        >
          Savings Account
        </td>
        <td
          class="px-2 text-right"
        >
          ₹50.00
        </td>
        <td
          class="px-2 text-right"
        >
          ₹500.00
        </td>
      </tr>
      <tr
        class="border-solid border-gray-400 border-y bg-sky-100"
      >
        <td
          class="px-2 py-1.5 pl-4"
        >
          Total Assets
        </td>
        <td
          class="px-2 text-right"
        >
          ₹200.00
        </td>
        <td
          class="px-2 text-right"
        >
          ₹720.00
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>"
`;
